### PR TITLE
fix nd array bugs

### PR DESCRIPTION
**Pull request by luke-jiang**

### CODE REVIEW

**Review by Harjas12 — APPROVED** [review]
Fixed the issue with 2d array being a pointer

File: src/main/scala/lms/thirdparty/cuda/cuda.scala [hunk 1]
```diff
@@ -442,10 +444,32 @@ trait CudaOps extends Dsl with StackArrayOps with SizeTOps with CLibs with CudaF
...
+  case class NdArray[T:Manifest](x: Rep[Array[T]], xs: Seq[Int]) {
+    val dim = xs.length
+    def offset(is: Seq[Rep[Int]]) = (xs, is).zipped.map(_*_).reduce((x, y) => x + y)
+    
+    def apply(is: Rep[Int]*)(implicit __pos: SourceContext): Rep[T] = {
+      assert(is.size == dim, "wrong arguments to N-d array")
+      Wrap[T](Adapter.g.reflectRead("array_get", Unwrap(x), Unwrap(offset(is)))(Unwrap(x)))
+    }
+    // (luke) TODO: what is a better way to do this?
+    def update(is: Rep[Int]*)(y: Rep[T])(implicit __pos: SourceContext): Unit = {
```
Comment (luke-jiang) [inline src/main/scala/lms/thirdparty/cuda/cuda.scala:461]: how to use sugared syntax for update?

File: src/test/scala/lms/thirdparty/test_cuda.scala [hunk 2]
```diff
@@ -243,7 +245,53 @@ class CudaTest extends TutorialFunSuite {
...
+        printf("%d", c(0)(1)(0))*/
+        val a = NewSharedArray[Int](1)
+        a(0) = 0
+        val b = NewSharedMatrix[Int](2, 2)
+        b(0, 0) = arg
+        printf("%d", b(0, 1))
+        printf("%d", b(1, 1))
+        
+        val c = NewSharedNdArray[Int](3, 2)
+        c.update(0, 0)(0)
```
Comment (luke-jiang) [inline src/test/scala/lms/thirdparty/test_cuda.scala:257]: should be `c(0, 0) = 0`

**Review by supunab — APPROVED** [review]
LGTM!

File: src/main/scala/lms/thirdparty/cuda/cuda.scala [hunk 3]
```diff
@@ -661,8 +681,10 @@ trait CudaOps extends Dsl with StackArrayOps with SizeTOps with CLibs with CudaF
 
 
   // Some global values for our GPU
+  val tileDim = 32
```
Comment (feiwang3311) [inline src/main/scala/lms/thirdparty/cuda/cuda.scala:684]: let's move this val to later. it is not global values of GPU

File: src/main/scala/lms/thirdparty/cuda/cuda.scala [hunk 4]
```diff
@@ -661,8 +681,10 @@ trait CudaOps extends Dsl with StackArrayOps with SizeTOps with CLibs with CudaF
...
   val gridSize = 28
   val blockSize = 512
+  val blockRows = 8
```
Comment (feiwang3311) [inline src/main/scala/lms/thirdparty/cuda/cuda.scala:687]: same here

File: src/main/scala/lms/thirdparty/cuda/cuda.scala [hunk 5]
```diff
@@ -812,6 +833,28 @@ trait CudaOps extends Dsl with StackArrayOps with SizeTOps with CLibs with CudaF
       }
     }
 
+  def cudaTranspose[N:Numeric:Manifest](implicit __pos: SourceContext) = cudaGlobalFun {
+    (in: Rep[Array[N]], out: Rep[Array[N]]) =>
+      val tile = NewSharedArray[N](tileDim, tileDim + 1)
```
Comment (feiwang3311) [inline src/main/scala/lms/thirdparty/cuda/cuda.scala:838]: need comments before here

File: src/main/scala/lms/thirdparty/cuda/cuda.scala [hunk 6]
```diff
@@ -994,6 +1037,10 @@ trait CCodeGenCudaOps extends CCodeGenSizeTOps with CudaCodeGenLibFunction with
 
   override def traverse(n: Node): Unit = n match {
     case n @ Node(s, "NewSharedArray", xs, _) =>
+      // var tpe = typeMap.get(s)
+      // (1 to xs.length) foreach { _ => tpe = tpe.map(_.typeArguments.head) }
+      // val tpeAfter = remap(tpe.getOrElse(manifest[Unknown]))
+      // emit("__shared__ "); emit(s"$tpeAfter "); shallow(s);
```
Comment (feiwang3311) [inline src/main/scala/lms/thirdparty/cuda/cuda.scala:1043]: remove commented code if we don't need them any more

File: src/test/scala/lms/thirdparty/test_cuda.scala [hunk 7]
```diff
@@ -235,12 +235,26 @@ class CudaTest extends TutorialFunSuite {
...
+        printf("%d", c(0, 1, 2))
+        printf("%d", c(1, 0, 0))
+        printf("%d", c(1, 0, 1))
+        printf("%d", c(1, 0, 2))
+        printf("%d", c(1, 1, 0))
+        printf("%d", c(1, 1, 1))
+        printf("%d", c(1, 1, 2))
       }
     }
     System.out.println(indent(driver.code))
```
Comment (feiwang3311) [inline src/test/scala/lms/thirdparty/test_cuda.scala:260]: use check file here?

File: src/test/scala/lms/thirdparty/test_cuda.scala [hunk 8]
```diff
@@ -311,7 +325,7 @@ class CudaTest extends TutorialFunSuite {
         }
       }
     }
-//    check("softmax", driver.code, "cu")
+    // check("softmax", driver.code, "cu")
```
Comment (feiwang3311) [inline src/test/scala/lms/thirdparty/test_cuda.scala:328]: use check file here?

File: src/test/scala/lms/thirdparty/test_cuda.scala [hunk 9]
```diff
@@ -376,5 +390,24 @@ class CudaTest extends TutorialFunSuite {
...
+
+        val transposeFun = cudaTranspose[Float]
+        transposeFun(in, out, dim3(32, 32, 1), dim3(2, 2, 1))
+
+        val out2 = NewArray[Float](a * a)
+        cudaCall(cudaMemcpyOfT[Float](out2, out, a * a, device2host))
+        printf("%f", out2(0))
+      }
+    }
+    System.out.println(indent(driver.code))
```
Comment (feiwang3311) [inline src/test/scala/lms/thirdparty/test_cuda.scala:410]: use check file here and also add pytorch test?

**Review by feiwang3311 — COMMENTED** [review]
good progress!

File: src/main/scala/lms/thirdparty/cuda/cuda.scala (hunk 6 above)
Comment (luke-jiang) [reply src/main/scala/lms/thirdparty/cuda/cuda.scala]: fixed in b61827c

File: src/test/scala/lms/thirdparty/test_cuda.scala (hunk 7 above)
Comment (luke-jiang) [reply src/test/scala/lms/thirdparty/test_cuda.scala]: fixed in 0b67e1e

File: src/test/scala/lms/thirdparty/test_cuda.scala (hunk 8 above)
Comment (supunab) [reply src/test/scala/lms/thirdparty/test_cuda.scala:329]: Yes, I need to implement the PyTorch tests as you did with embedding

**Pull request merged**